rfcs/20180827-api-names.md

```diff
@@ -0,0 +1,607 @@
+*   To keep clear structure and reduce duplication, we want to collect all layers, losses and metrics under the `tf.keras` namespace.
+
+In general, we want to balance flatness and browsability. Flat hierarchies allow
+for shorter endpoint names that are easy to remember (for e.g. `tf.add` vs
+`tf.math.add`). At the same time subnamespaces support easier browsability (for
+e.g. `tf.math` namespace would contain all math functions making it easier to discover available symbols).
+
+Furthermore, TensorFlow API has many users. Therefore, we should avoid removing endpoints if they are frequently used.
+
+## Design Proposal
+
+## Additional namespaces
+
+We plan to add the following additional namespaces:
+
+**tf.random** - will contain random sampling ops.
+**tf.keras.layers** - will contain all symbols that are currently under `tf.layers`. Note that signatures of these symbols will likely change to match layers under tf.keras.layers better.
+**tf.keras.losses** - will contain all symbols that are currently under `tf.losses`. Note that signatures of these symbols will likely change to match losses under tf.keras.losses better.
+**tf.keras.metrics** - will contain all symbols that are currently under `tf.metrics`. Note that signatures of these symbols will likely change to match metrics under tf.keras.metrics better.
+
+
+Note that we already introduced some new namespaces earlier in June, specifically
+
+**tf.debugging** - ops helpful for debugging, such as asserts. We also want to
+move [TensorFlow Debugger](https://www.tensorflow.org/guide/debugger) to
+`tf.debugging` namespace.
+
+**tf.dtypes** - data types.
+
+**tf.io** - ops for reading and writing.
+
+**tf.quantization** - ops related to quantization.
+
+## Deprecated namespaces
+
+We plan to deprecate entire contents of the following namespaces:
+
+**tf.logging** - Python `logging` module can be used instead.
+
+
+## Additional endpoints
+
+We will add new endpoints for existing symbols to make sure each namespace contains all relevant endpoints.
+
+See list of endpoints we want to add in *Appendix 1: Additional
+Endpoints*. Note: the list in the appendix does not include new endpoints for symbols under
+`tf.layers`, `tf.losses` and `tf.metrics` namespaces since all symbols under
+these namespaces will have new endpoints added under `tf.keras.layers`,
+`tf.keras.losses` and `tf.keras.metrics` respectively. So, we don't need to
+list these endpoint changes individually.
+
+
+## Deprecated endpoints
+
+We also want to remove some of the existing endpoints. Specifically we were looking for endpoints to remove based on the following criteria:
+
+*   Remove endpoints if they have preferred replacement and if these endpoints are not frequently used.
+*   Remove all endpoints that have been moved to `tf.quantization` namespace.
+*   Remove all endpoints that have been moved to `tf.random` namespace.
+*   Remove all endpoints from `tf.logging`.
+*   Keeps most symbols in `tf.manip` in root namespace.
```

Comment:
Should we bother making manip? It's a mixture of set_ops and array_ops, and since we're keeping all of them (most of them?) in the root anyway, how about this:

tf.sets.* (only) for the set ops currently in manip
tf.* (only) for the tensor manipulation ops currently in manip

?

Reply:
There is also manip_ops:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/manip_ops.cc
It currently only has 1 op (tf.manip.roll).
We also have tf.manip already:
https://www.tensorflow.org/api_docs/python/tf/manip

That being said, tf.manip.roll is not that frequently used and I think most references to other ops in tf.manip are still referring to these ops in root. So, we can probably deprecate tf.manip. It is just a bit strange since we just create it recently.

Reply:
+1 for not bothering with `tf.manipt`